File: frontend/coordinator-android/engine/src/main/kotlin/ug/qaat/engine/SessionManager.kt
package ug.qaat.engine

/**
 * Session lifecycle — port of apps/coordinator-pwa/src/session/state-machine.ts.
 * IDLE → PENDING_LECTURER → ACTIVE → CLOSING → CLOSED, with the same auto-expiry
 * rules (check-in window T+120m from gate-open; hard force-close T+180m from start).
 */
enum class SessionState { IDLE, PENDING_LECTURER, ACTIVE, CLOSING, CLOSED }

class SessionManager(
    val sessionId: String,
    private val nowMillis: () -> Long = { System.currentTimeMillis() },
) {
    var state: SessionState = SessionState.IDLE; private set
    var startedAtMs: Long = 0; private set
    var gateOpenAtMs: Long = 0; private set
    var closedAtMs: Long = 0; private set

    companion object {
        const val CHECKIN_WINDOW_MIN = 120L
        const val FORCE_CLOSE_MIN = 180L
    }

    /** Coordinator opens the session; awaits the lecturer gate. */
    fun open() {
        require(state == SessionState.IDLE) { "session already started" }
        state = SessionState.PENDING_LECTURER
        startedAtMs = nowMillis()
    }

    /** Lecturer passes the START gate → check-in window opens. */
    fun lecturerStarted() {
        require(state == SessionState.PENDING_LECTURER) { "not awaiting lecturer" }
        state = SessionState.ACTIVE
        gateOpenAtMs = nowMillis()
    }

    /** Whether students may currently check in (active + inside the T+120 window). */
    fun checkinOpen(): Boolean =
        state == SessionState.ACTIVE && nowMillis() < gateOpenAtMs + CHECKIN_WINDOW_MIN * 60_000

    fun beginClose() { if (state == SessionState.ACTIVE) state = SessionState.CLOSING }

    fun close() { state = SessionState.CLOSED; closedAtMs = nowMillis() }

    /**
     * Apply auto-expiry; returns true if the session was force-closed.
     * Hard kill at T+180m from start regardless of coordinator action.
     */
    fun tickAutoExpiry(): Boolean {
        if (state == SessionState.CLOSED) return false
        if (startedAtMs > 0 && nowMillis() >= startedAtMs + FORCE_CLOSE_MIN * 60_000) {
            close(); return true
        }
        return false
    }
}

File: frontend/coordinator-flutter/lib/engine/session_manager.dart
enum SessionState { idle, pendingLecturer, active, closing, closed }

class SessionManager {
  final String sessionId;
  final int Function() nowMillis;

  SessionManager(this.sessionId, {int Function()? nowMillis})
      : nowMillis = nowMillis ?? _defaultNowMillis;

  static int _defaultNowMillis() => DateTime.now().millisecondsSinceEpoch;

  static const int checkinWindowMin = 120;
  static const int forceCloseMin = 180;

  SessionState _state = SessionState.idle;
  int _startedAtMs = 0;
  int _gateOpenAtMs = 0;
  int _closedAtMs = 0;

  SessionState get state => _state;
  int get startedAtMs => _startedAtMs;
  int get gateOpenAtMs => _gateOpenAtMs;
  int get closedAtMs => _closedAtMs;

  void open() {
    if (_state != SessionState.idle) {
      throw StateError('session already started');
    }
    _state = SessionState.pendingLecturer;
    _startedAtMs = nowMillis();
  }

  void lecturerStarted() {
    if (_state != SessionState.pendingLecturer) {
      throw StateError('not awaiting lecturer');
    }
    _state = SessionState.active;
    _gateOpenAtMs = nowMillis();
  }

  bool checkinOpen() =>
      _state == SessionState.active &&
      nowMillis() < _gateOpenAtMs + checkinWindowMin * 60000;

  void beginClose() {
    if (_state == SessionState.active) {
      _state = SessionState.closing;
    }
  }

  void close() {
    _state = SessionState.closed;
    _closedAtMs = nowMillis();
  }

  bool tickAutoExpiry() {
    if (_state == SessionState.closed) return false;
    if (_startedAtMs > 0 && nowMillis() >= _startedAtMs + forceCloseMin * 60000) {
      close();
      return true;
    }
    return false;
  }
}